### PR TITLE
Implement basic no-op reconciler for all Runs

### PR DESCRIPTION
**Pull request by ImJasonH**
This wires up a reconciler in the core controller binary that is
notified of changes to and creations of all Run objects of any type. It
currently does nothing in response to these events, but in the future
this will be used to enforce initial update timeouts, and possibly more
default behavior in the future.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [n] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [y] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [n] Release notes block has been filled in or deleted (only if no user facing changes)

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: cmd/controller/main.go [hunk 1]
```diff
@@ -74,5 +75,6 @@ func main() {
 	sharedmain.MainWithContext(injection.WithNamespaceScope(signals.NewContext(), *namespace), ControllerLogKey,
 		taskrun.NewController(*namespace, images),
 		pipelinerun.NewController(*namespace, images),
+		run.NewController,
```
Comment (pritidesai) [inline cmd/controller/main.go:78]: sorry coming late to the party 😢 but staring at this `run` here looks incomplete, is it possible to rename it to something descriptive? 😜

File: pkg/reconciler/run/controller.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	runreconciler "github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/tracker"
+)
+
+// NewController instantiates a new controller.Impl from knative.dev/pkg/controller
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
```
Comment (pritidesai) [inline pkg/reconciler/run/controller.go:35]: I am still trying to understand how controller initialization is happening in general. Here `NewController` is returning   implementation of the controller whereas `taskRun` and `pipelineRun` controllers are returning function:

https://github.com/tektoncd/pipeline/blob/732d2363b9df0b5c2e6b1de732a819b69c70bb3f/pkg/reconciler/taskrun/controller.go#L45-L46

https://github.com/tektoncd/pipeline/blob/732d2363b9df0b5c2e6b1de732a819b69c70bb3f/pkg/reconciler/pipelinerun/controller.go#L45-L46

Why does this defer from the `pipelineRun` and `taskRun`?

File: pkg/reconciler/run/run.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2019 The Tekton Authors
```
Comment (pritidesai) [inline pkg/reconciler/run/run.go:2]: may be `2020` similar to one in `controller.go` ⬆️ ?

File: pkg/reconciler/run/run.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+)
+
+const (
+	// runAgentName defines logging agent name for Run Controller
+	runAgentName = "run-controller"
+)
+
+// Reconciler implements controller.Reconciler for Run resources.
+type Reconciler struct {
+	pipelineClientSet clientset.Interface
```
Comment (pritidesai) [inline pkg/reconciler/run/run.go:37]: `s/pipelineClientSet/runClientSet/` 🤔

File: pkg/reconciler/run/run.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	tracker           tracker.Interface
+}
+
+// Check that our Reconciler implements runreconciler.Interface
+var _ runreconciler.Interface = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Task Run
```
Comment (pritidesai) [inline pkg/reconciler/run/run.go:45]: s/Task Run/Run/

File: pkg/reconciler/run/controller.go (hunk 2 above)
Comment (vdemeester) [reply pkg/reconciler/run/controller.go:35]: Good point. I think we may want to follow the same pattern. In PipelineRun and TaskRun reconciler, this is mainly because we need to pass arguments (and the func return to satisfy the knative/pkg controller contract). We should support at least the `namespace` for `Run` too I guess ? (I don't think we need to pass the `images`)

File: cmd/controller/main.go (hunk 1 above)
Comment (ImJasonH) [reply cmd/controller/main.go:78]: The `run` package is generated based on the name of the type, which is just `Run`. I can't really think of a better name for it, but ideas are welcome. An earlier iteration of the design called it `CustomTaskRun` but that got to be a mouthful/keyboardful...

File: pkg/reconciler/run/controller.go (hunk 2 above)
Comment (ImJasonH) [reply pkg/reconciler/run/controller.go:35]: I had considered supporting namespaced-reconciliation for Runs in this PR but thought it would be simpler not to, to keep this change small. If you think it makes it easier to read here, though, I can do that.

File: pkg/reconciler/run/run.go (hunk 4 above)
Comment (ImJasonH) [reply pkg/reconciler/run/run.go:37]: This is actually named `pipeline` because it's the "Tekton Pipelines" client -- it provides (generated) methods to interact with _all_ "Tekton Pipelines" resources and API versions: https://godoc.org/github.com/tektoncd/pipeline/pkg/client/clientset/versioned

It's also named `pipelineClientSet` in the TaskRun controller too, for example: https://github.com/tektoncd/pipeline/blob/732d2363b9df0b5c2e6b1de732a819b69c70bb3f/pkg/reconciler/taskrun/taskrun.go#L64

**Comment by ImJasonH** [comment]
```
==================================
==== INTEGRATION TESTS PASSED ====
==================================
```
/retest 
🤔

File: cmd/controller/main.go (hunk 1 above)
Comment (pritidesai) [reply cmd/controller/main.go:78]: is it distinguishable as a `Task` or a `Pipeline`?  How about `CustomRun`? 😜

File: pkg/reconciler/run/controller.go (hunk 2 above)
Comment (pritidesai) [reply pkg/reconciler/run/controller.go:35]: Yup would be great to exercise the same contract and easier to follow. If it cannot be part of this PR, at least a comment would be helpful 🙏

File: pkg/reconciler/run/run.go (hunk 4 above)
Comment (pritidesai) [reply pkg/reconciler/run/run.go:37]: thanks for the explanation 👍

File: cmd/controller/main.go (hunk 1 above)
Comment (ImJasonH) [reply cmd/controller/main.go:78]: The general idea is that a `Run` reports the execution of a generic...something. What that is is up to the custom task author -- it could be running a remote build process, "running" a wait between two other Tasks, "running" an approval.

It's sort of like a generic container to express an ongoing execution of a thing. In the future we could experiment with having `Run<Task>` and `Run<Pipeline>` in addition to (eventually replacing?) `TaskRun` and `PipelineRun`. This would give us flexibility to experiment with new definitions of `Task` and `Pipeline` outside the `tektoncd/pipeline` repo, or to mix versions of these types (e.g., `Run<v1alpha1.Task>` alongside `Run<v1.Task>` in the same `PipelineRun`).

So I think even `CustomRun` is still a bit wordy for what it could end up supporting. What's a "custom" run in a possible future world where that's actually the _only_ kind of running thing? I personally actually kind of like the succintness of "`Run`"  in that context. 🤷‍♂️

Comment (pritidesai) [reply cmd/controller/main.go:78]: cool, sounds good, thanks 👍

File: pkg/reconciler/run/controller.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+	logger.Info("Setting up event handlers")
+	runInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    impl.Enqueue,
+		UpdateFunc: controller.PassNew(impl.Enqueue),
+	})
+	c.tracker = tracker.New(impl.EnqueueKey, controller.GetTrackerLease(ctx))
+
+	return impl
+}
```
Comment (bobcatfish) [inline pkg/reconciler/run/controller.go:61]: if the intention of this controller is for folks to take it and build on it (I think? side note im not quite sure what this controller is for from the PR description / commit message) then maybe it would make sense to include some comments for every few lines explaining whats going on?

File: go.mod [hunk 7]
```diff
@@ -3,6 +3,7 @@ module github.com/tektoncd/pipeline
 go 1.13
 
 require (
+	cloud.google.com/go/storage v1.6.0
```
Comment (dlorenc) [inline go.mod:6]: Any idea why this was added? It doesn't appear used anywhere.

Comment (vdemeester) [reply go.mod:6]: I gets added and removed as will (or random :sweat_smile:). That is one of the *high* mistery of go modules with k8s I feel :sweat_smile:

Comment (pritidesai) [reply go.mod:6]: no idea but leaves us with `git checkout -- go.mod` 😢

**Comment by vdemeester** [comment]
@ImJasonH what is the status for this ?

**Comment by ImJasonH** [comment]
I think there were questions about whether we wanted to have Tekton reconcile anything about Runs, which leaves this in limbo.

I'm fine to close now and reopen (or something like it) if we decide to have Tekton reconcile anything.

/close
